### DISCUSSION TITLE
HTTP cache control

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#95] [프론트 기술 질문 공간]
1 . http 통신의  cache control 헤더에 올 수 있는 값들을 설명해주세요 

2. 캐싱된 데이터에 대한 validation을 위한 값은 무엇이 있고 어떻게 사용하는지 설명해주세요

**Comment by michoball** [comment]
cache-control 값

no-store:  캐시를 저장하지 않고 매번 서버로부터 데이터를 다운 받는다

no-cache: 캐시를 하지만 요청마다 캐시 유효성 검사를 한다

public, private: public은 공유 캐시나 중개 서버에 캐시를 저장해도 된다는 뜻 
private 는 브라우져같은 특정 사용자 환경에만 저장

max-age: 여기에 쓰여진 seconds 시간만큼은 캐시가 유효하다는 것을 뜻한다.

must-revalidate: 만료된 캐시는 검증을 해야한다는 것을 뜻한다.

캐시 유효성검사

실효된 리소스 (stale)에 대한 요청을 받은 경우 유효한지 확인을 위해 If-None-Match 헤더를 포함해서 요청을 전달합니다. 그럼 서버는 리소스 본문을 전송하지 않고 304 헤더를 돌려보냅니다. 

max-age 헤더가 있으면 거기에 있는 시간이 유효수명, Expires 헤더가 있으면 거시서 Date 헤더 값을 뺀 결과가 유효수명 입니다.

따로 명시적으로 유효성 헤더를 지정하지 않으면 Last-Modified 헤더를 찾아서 
Date 헤더 값에서 Last-Modified 값을 빼고 10으로 나눈결과가 유효수명입니다. 

캐시 검증

문서의 만료 시간이 가까워져오면 검증하거나 다시 리소스를 불러옵니다. 강한점증과 약한 검증 중 하나라도  제공하는 경우에 일어납니다.

ETags  강한 검증으로 사용될 수 있는 값입니다.  이 헤더가 응답에 있으면 클라이언트는 이후 요청의 헤더내에 If-None-Match를 줄 수 있고 이 헤더를 보고 ETag와 값이 다르면 새로 리소스를 가져오게 합니다.  

 Last-Modified  약한 검증으로 사용될 수 있습니다.  Last-Modified 가 응답에 있으면 클라이언트는 검증을 위해 If-Modified-Since 를 줄 수 있고 이걸보고 검증을 합니다. 

검증을 하면 200 번으로 새로 리소스를 주거나 304 번을 반환합니다.

**Comment by WooGie911** [comment]
1 . http 통신의 cache control 헤더에 올 수 있는 값들을 설명해주세요

HTTP/1.1의 Cache-Control 헤더에는 HTTP 요청/응답에서의 캐싱 메커니즘을 결정하는 여러 디렉티브(Directive)들을 나열할 수 있습니다. 
각각의 디렉티브는 캐싱을 어떻게 할 것인지와 관련된 일종의 옵션이라고 보면 됩니다. 
이때 Cache-Control 헤더는 HTTP 요청 헤더와 HTTP 응답 헤더에 모두 사용할 수 있지만, 각각에 나열하는 디렉티브들은 서로 다른 의미를 지니며 나열할 수 있는 디렉티브들의 종류도 조금씩 다릅니다. 
따라서 HTTP 요청의 Cache-Control 헤더에 나열된 디렉티브들이 반드시 HTTP 응답의 Cache-Control 헤더에 나열되리라는 보장은 없습니다. 
디렉티브 이름은 대소문자를 구별하지 않으며, 여러 디렉티브들을 나열하는 경우에는 콤마로 구분합니다.

![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/825cdfdc-1505-4bca-a19d-25d68e85d5d8)

2. 캐싱된 데이터에 대한 validation을 위한 값은 무엇이 있고 어떻게 사용하는지 설명해주세요

만약 수명이 아직 다하지 않았다면 해당 HTTP 응답은 유효하다(Fresh)고 표현하며, 수명이 다했다면 만료되었다(Stale)고 표현합니다. 
기본적으로 캐시는 만료된 자원에 대한 HTTP 요청을 받으면 그 자원이 여전히 유효한지 검증하기 위해 서버에게 해당 HTTP 요청을 전달합니다. 
이러한 과정을 검증(Validation)이라고 하며, 검증하고자 하는 HTTP 응답의 헤더 구성에 따라 검증 요청 시 서버에 전달하는 정보의 종류가 조금 달라집니다. 

검증 방식은 크게 두 가지입니다. 

1. 검증하고자 하는 HTTP 응답에 ETag 헤더가 존재한다면 ETag 헤더의 값을 If-None-Match 헤더에 포함시켜서 서버에게 검증 요청을 보냅니다. 

2. ETag 헤더가 존재하지 않는다면 Last-Modified 헤더를 활용합니다. 
즉, Last-Modified 헤더의 값을 If-Modified-Since 헤더에 포함시켜서 서버에게 검증 요청을 보냅니다.

이제 서버는 전달받은 ETag 값 혹은 Last-Modified 값을 이용하여 자원의 유효성을 검증합니다. 
만약 여전히 해당 자원이 유효하다면 서버는 Body가 없는 가벼운 304 (Not Modified) 응답을 반환하고, 더 이상 유효하지 않다면 새로운 자원의 내용을 Body에 담아서 200 응답을 반환합니다. 
전자의 경우에는 기존에 캐싱되어 있던 자원을 다시 사용하게 하는 것이므로 나이(Age)를 0으로 초기화시키게 되고, 필요한 경우에는 만료 시각을 갱신해줄 수도 있습니다.
